File: local/python/ncp_local/modular_client.py
"""Independent one-pending client. Unknown channel outcomes never grant retry."""
from __future__ import annotations

from typing import BinaryIO

from . import wire as framing
from .modular_buffer import BufferBinding, MAX_ID
from .modular_owner import profile_digest, verify_response, verify_retrieved
from .modular_wire import (Ack, Command, Contract, Execute, ModularError, Name, Operation,
                           Outcome, Query, Request, Response, PROFILE_DOMAIN, operation_name, parse, typed_digest)

CLIENT_LOGICAL_OVERHEAD = 1_085_440


class Client:
    """Six frame extents, three parsed extents, explicit UTF8 scratch, and fixed state.

    The logical allowance includes one caller-held typed original during verification.
    Extra caller-retained copies and installed application objects need separate bounds.
    This is not an allocator or process RSS guarantee.
    """
    def __init__(self, binding: BufferBinding, contract: type[Contract]) -> None:
        binding.validate()
        if binding.profile_digest != profile_digest() or binding.application_digest != typed_digest(PROFILE_DOMAIN, parse(contract.descriptor())):
            raise ModularError("binding")
        self._binding = binding
        self._contract = contract
        self._next_sequence = 1
        self._predecessor = None
        self._pending = None
        self._result_digest = None
        self._retired = False

    @property
    def binding(self) -> BufferBinding: return self._binding
    @property
    def next_sequence(self) -> int | None: return self._next_sequence
    @property
    def predecessor(self) -> str | None: return self._predecessor
    @property
    def pending_request(self) -> bytes | None: return self._pending
    @property
    def is_retired(self) -> bool: return self._retired

    def retire_channel(self) -> None: self._retired = True

    def begin(self, operation: Operation) -> bytes:
        """Complete local preflight and immutable journal bytes before any write."""
        if self._retired: raise ModularError("retired")
        if self._pending is not None: raise ModularError("binding")
        if self._next_sequence is None: raise ModularError("capacity")
        self._contract.check_input(operation)
        if not self._contract.allows(operation_name(operation)): raise ModularError("binding")
        payload = Request.create(self.binding, self._next_sequence, Execute(self._predecessor, operation))
        parsed = Request.decode(payload, self.binding, self._contract)
        self._contract.check_input(parsed.command.operation)
        self._pending = payload
        return payload

    def _original(self) -> Request:
        if self._retired: raise ModularError("retired")
        if self._pending is None: raise ModularError("binding")
        return Request.decode(self._pending, self.binding, self._contract)

    def result_query(self) -> bytes:
        original = self._original()
        return Request.create(self.binding, original.sequence, Query(original.request_digest))

    def acknowledgement(self) -> bytes:
        original = self._original()
        if self._result_digest is None: raise ModularError("binding")
        return Request.create(self.binding, original.sequence, Ack(original.request_digest, self._result_digest))

    def _apply(self, response: Response) -> None:
        if self._result_digest is not None and (response.outcome != Outcome.COMMITTED or response.result_digest != self._result_digest):
            raise ModularError("binding")
        if response.outcome == Outcome.COMMITTED: self._result_digest = response.result_digest
        elif response.outcome == Outcome.REJECTED:
            self._pending = None
            self._result_digest = None
        elif response.outcome == Outcome.INDETERMINATE:
            self._result_digest = response.result_digest
            self.retire_channel()
        elif response.outcome not in (Outcome.NOT_ADMITTED, Outcome.UNAVAILABLE): raise ModularError("wire")

    def observe(self, payload: bytes | memoryview) -> Response:
        try:
            original = self._original()
            response = verify_response(self.binding, original, payload, self._contract)
            self._apply(response)
            return response
        except BaseException:
            self.retire_channel()
            raise

    def observe_query(self, payload: bytes | memoryview) -> Response:
        try:
            original = self._original()
            query = Request.decode(self.result_query(), self.binding, self._contract)
            shape = parse(payload)
            unavailable = shape.get("operation") == Name.RESULT.value
            del shape
            if unavailable: return verify_response(self.binding, query, payload, self._contract)
            response = verify_retrieved(self.binding, original, query, payload, self._contract)
            self._apply(response)
            return response
        except BaseException:
            self.retire_channel()
            raise

    def observe_acknowledgement(self, payload: bytes | memoryview) -> Response:
        try:
            ack = Request.decode(self.acknowledgement(), self.binding, self._contract)
            response = verify_response(self.binding, ack, payload, self._contract)
            if response.outcome == Outcome.ACKNOWLEDGED:
                self._predecessor = self._result_digest
                self._result_digest = None
                self._next_sequence = ack.sequence + 1 if ack.sequence < MAX_ID else None
                self._pending = None
            return response
        except BaseException:
            self.retire_channel()
            raise

    def dispatch(self, reader: BinaryIO, writer: BinaryIO, *, deadline: float) -> Response:
        """One explicit dispatch, with an absolute monotonic deadline and no retry."""
        if self._retired or self._pending is None or self._result_digest is not None: raise ModularError("binding")
        try:
            framing.write_local_frame(writer, self._pending, deadline=deadline)
            payload = framing.read_local_frame(reader, deadline=deadline)
            if payload is None: raise ModularError("retired")
            return self.observe(payload)
        except BaseException as error:
            self.retire_channel()
            raise ModularError("retired") from error

    def dispatch_acknowledgement(self, reader: BinaryIO, writer: BinaryIO, *, deadline: float) -> Response:
        payload = self.acknowledgement()
        try:
            framing.write_local_frame(writer, payload, deadline=deadline)
            response = framing.read_local_frame(reader, deadline=deadline)
            if response is None: raise ModularError("retired")
            return self.observe_acknowledgement(response)
        except BaseException as error:
            self.retire_channel()
            raise ModularError("retired") from error

    def dispatch_query(self, reader: BinaryIO, writer: BinaryIO, *, deadline: float) -> Response:
        payload = self.result_query()
        try:
            framing.write_local_frame(writer, payload, deadline=deadline)
            response = framing.read_local_frame(reader, deadline=deadline)
            if response is None: raise ModularError("retired")
            return self.observe_query(response)
        except BaseException as error:
            self.retire_channel()
            raise ModularError("retired") from error

File: ncp-core/src/modular_client.rs
//! One-pending-request client. A channel failure preserves evidence and forbids retry.
//! Pure response verification is available separately in `modular_owner`.

use std::io::{Read, Write};
use std::marker::PhantomData;
use std::panic::{catch_unwind, AssertUnwindSafe};

use crate::modular_buffer::{BufferBinding, MAX_ID};
use crate::modular_owner::{self, AppOperation, AppResponse, Contract};
use crate::modular_wire::{self as wire, Command, ModularError, Outcome, Request};

pub const CLIENT_LOGICAL_OVERHEAD: usize = 1_085_440;

/// Logical storage includes six bounded frame extents, three parsed extents,
/// digest/sort, decoded-chunk and explicit UTF8 scratch, and fixed identity state.
/// These are logical extents, not an allocator, application-object, or RSS bound.
/// A caller-held typed original occupies one parsed extent during verification.
/// Independently retained originals, responses, and artifacts need caller accounting.
pub struct Client<A: Contract> {
    binding: BufferBinding,
    next_sequence: Option<u64>,
    predecessor: Option<String>,
    pending: Vec<u8>,
    result_digest: Option<String>,
    retired: bool,
    contract: PhantomData<A>,
}

impl<A: Contract> Client<A> {
    pub fn new(binding: BufferBinding) -> Result<Self, ModularError> {
        binding.validate().map_err(|_| ModularError::Binding)?;
        let application = wire::typed_digest(
            wire::PROFILE_DOMAIN,
            &wire::parse_value(A::descriptor())?,
            None,
        )?;
        if binding.profile_digest != modular_owner::profile_digest()?
            || binding.application_digest != application
        {
            return Err(ModularError::Binding);
        }
        Ok(Self {
            binding,
            next_sequence: Some(1),
            predecessor: None,
            pending: wire::frame_slot()?,
            result_digest: None,
            retired: false,
            contract: PhantomData,
        })
    }

    pub fn binding(&self) -> &BufferBinding {
        &self.binding
    }
    pub fn next_sequence(&self) -> Option<u64> {
        self.next_sequence
    }
    pub fn predecessor(&self) -> Option<&str> {
        self.predecessor.as_deref()
    }
    pub fn pending_request(&self) -> Option<&[u8]> {
        (!self.pending.is_empty()).then_some(self.pending.as_slice())
    }
    pub fn is_retired(&self) -> bool {
        self.retired
    }
    pub fn retire_channel(&mut self) {
        self.retired = true;
    }

    /// Local preflight only. No channel write occurs, and failure leaves state intact.
    /// The returned immutable bytes can be durably journaled before dispatch.
    pub fn begin(&mut self, operation: AppOperation<A>) -> Result<&[u8], ModularError> {
        if self.retired {
            return Err(ModularError::Retired);
        }
        if !self.pending.is_empty() {
            return Err(ModularError::Binding);
        }
        let sequence = self.next_sequence.ok_or(ModularError::Capacity)?;
        A::check_input(&operation)?;
        if !A::allows(operation.name()) {
            return Err(ModularError::Binding);
        }
        let bytes = Request::encode(
            self.binding.clone(),
            sequence,
            Command::Execute {
                expected_predecessor_result_digest: self.predecessor.clone(),
                operation,
            },
        )?;
        // Reopen the exact representation before publishing pending state.
        let parsed = Request::<AppOperation<A>>::decode(&bytes, &self.binding)?;
        let Command::Execute { operation, .. } = &parsed.command else {
            return Err(ModularError::Wire);
        };
        A::check_input(operation)?;
        drop(parsed);
        self.pending.extend_from_slice(&bytes);
        Ok(&self.pending)
    }

    fn original(&self) -> Result<Request<AppOperation<A>>, ModularError> {
        if self.retired {
            return Err(ModularError::Retired);
        }
        Request::decode(&self.pending, &self.binding)
    }

    pub fn result_query(&self) -> Result<Vec<u8>, ModularError> {
        let original = self.original()?;
        Request::<AppOperation<A>>::encode(
            self.binding.clone(),
            original.sequence,
            Command::Result {
                original_request_digest: original.request_digest,
            },
        )
    }

    pub fn acknowledgement(&self) -> Result<Vec<u8>, ModularError> {
        let original = self.original()?;
        let result_digest = self.result_digest.clone().ok_or(ModularError::Binding)?;
        Request::<AppOperation<A>>::encode(
            self.binding.clone(),
            original.sequence,
            Command::Ack {
                original_request_digest: original.request_digest,
                result_digest,
            },
        )
    }

    /// Accept only the exact original outcome. Invalid peer bytes retire the client.
    pub fn observe(&mut self, bytes: &[u8]) -> Result<AppResponse<A>, ModularError> {
        let result = (|| {
            let original = self.original()?;
            let response = modular_owner::verify_response::<A>(&self.binding, &original, bytes)?;
            self.apply_outcome(&response)?;
            Ok(response)
        })();
        if result.is_err() {
            self.retire_channel();
        }
        result
    }

    pub fn observe_query(&mut self, bytes: &[u8]) -> Result<AppResponse<A>, ModularError> {
        let result = (|| {
            let original = self.original()?;
            let query = Request::<AppOperation<A>>::decode(&self.result_query()?, &self.binding)?;
            // Unavailable is correlated to the query. Retained bytes belong to the original.
            let shape = wire::parse_value(bytes)?;
            let unavailable =
                shape.get("operation").and_then(serde_json::Value::as_str) == Some("result");
            drop(shape);
            if unavailable {
                return modular_owner::verify_response::<A>(&self.binding, &query, bytes);
            }
            let response =
                modular_owner::verify_retrieved::<A>(&self.binding, &original, &query, bytes)?;
            self.apply_outcome(&response)?;
            Ok(response)
        })();
        if result.is_err() {
            self.retire_channel();
        }
        result
    }

    fn apply_outcome(&mut self, response: &AppResponse<A>) -> Result<(), ModularError> {
        if let Some(known) = &self.result_digest {
            if response.outcome != Outcome::Committed || known != &response.result_digest {
                return Err(ModularError::Binding);
            }
        }
        match response.outcome {
            Outcome::Committed => self.result_digest = Some(response.result_digest.clone()),
            Outcome::RejectedBeforeExecution => {
                self.pending.clear();
                self.result_digest = None;
            }
            Outcome::Indeterminate => {
                self.result_digest = Some(response.result_digest.clone());
                self.retire_channel();
            }
            Outcome::NotAdmitted | Outcome::Unavailable => {}
            Outcome::Acknowledged => return Err(ModularError::Wire),
        }
        Ok(())
    }

    /// Only a verified ACK advances sequence/head and clears a committed request.
    pub fn observe_acknowledgement(
        &mut self,
        bytes: &[u8],
    ) -> Result<AppResponse<A>, ModularError> {
        let result = (|| {
            let ack = Request::<AppOperation<A>>::decode(&self.acknowledgement()?, &self.binding)?;
            let response = modular_owner::verify_response::<A>(&self.binding, &ack, bytes)?;
            if response.outcome == Outcome::Acknowledged {
                self.predecessor = self.result_digest.take();
                self.next_sequence = (ack.sequence < MAX_ID).then(|| ack.sequence + 1);
                self.pending.clear();
            }
            Ok(response)
        })();
        if result.is_err() {
            self.retire_channel();
        }
        result
    }

    /// Dispatch a prepared request exactly once. The host supplies deadlines.
    pub fn dispatch<R: Read, W: Write>(
        &mut self,
        reader: &mut R,
        writer: &mut W,
    ) -> Result<AppResponse<A>, ModularError> {
        if self.retired || self.pending.is_empty() || self.result_digest.is_some() {
            return Err(ModularError::Binding);
        }
        let result = catch_unwind(AssertUnwindSafe(|| {
            crate::local::write_local_frame(writer, &self.pending)
                .map_err(|_| ModularError::Retired)?;
            writer.flush().map_err(|_| ModularError::Retired)?;
            let bytes = crate::local::read_local_frame(reader)
                .map_err(|_| ModularError::Retired)?
                .ok_or(ModularError::Retired)?;
            self.observe(&bytes)
        }))
        .unwrap_or(Err(ModularError::Retired));
        if result.is_err() {
            self.retire_channel();
        }
        result
    }

    /// A failed ACK exchange retires the channel and retains unresolved pending bytes.
    pub fn dispatch_acknowledgement<R: Read, W: Write>(
        &mut self,
        reader: &mut R,
        writer: &mut W,
    ) -> Result<AppResponse<A>, ModularError> {
        let ack = self.acknowledgement()?;
        let result = catch_unwind(AssertUnwindSafe(|| {
            crate::local::write_local_frame(writer, &ack).map_err(|_| ModularError::Retired)?;
            writer.flush().map_err(|_| ModularError::Retired)?;
            let bytes = crate::local::read_local_frame(reader)
                .map_err(|_| ModularError::Retired)?
                .ok_or(ModularError::Retired)?;
            self.observe_acknowledgement(&bytes)
        }))
        .unwrap_or(Err(ModularError::Retired));
        if result.is_err() {
            self.retire_channel();
        }
        result
    }

    pub fn dispatch_query<R: Read, W: Write>(
        &mut self,
        reader: &mut R,
        writer: &mut W,
    ) -> Result<AppResponse<A>, ModularError> {
        let query = self.result_query()?;
        let result = catch_unwind(AssertUnwindSafe(|| {
            crate::local::write_local_frame(writer, &query).map_err(|_| ModularError::Retired)?;
            writer.flush().map_err(|_| ModularError::Retired)?;
            let bytes = crate::local::read_local_frame(reader)
                .map_err(|_| ModularError::Retired)?
                .ok_or(ModularError::Retired)?;
            self.observe_query(&bytes)
        }))
        .unwrap_or(Err(ModularError::Retired));
        if result.is_err() {
            self.retire_channel();
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modular_owner::{
        test_contract::{binding, TextApplication},
        Owner,
    };
    use std::cell::Cell;
    use std::rc::Rc;

    #[test]
    fn maximum_ack_enters_exhaustion_without_advancing_past_the_safe_integer() {
        let mut client = Client::<TextApplication>::new(binding()).unwrap();
        client.next_sequence = Some(MAX_ID);
        let pending = client
            .begin(AppOperation::<TextApplication>::Prepare(0))
            .unwrap()
            .to_vec();
        // A closed synthetic response models a caller that has already reached this boundary.
        let original =
            Request::<AppOperation<TextApplication>>::decode(&pending, &binding()).unwrap();
        let response = wire::Response::<String, wire::Empty, wire::Empty> {
            schema: wire::RESPONSE_SCHEMA.into(),
            binding: binding(),
            sequence: MAX_ID,
            operation: wire::OperationName::Prepare,
            request_digest: original.request_digest,
            outcome: Outcome::Committed,
            code: wire::Code::Ok,
            body: wire::Body::Prepared {
                data: String::new(),
            },
            result_digest: String::new(),
        };
        let mut bytes = wire::frame_slot().unwrap();
        wire::seal_into(response, "result_digest", wire::RESPONSE_SCHEMA, &mut bytes).unwrap();
        client.observe(&bytes).unwrap();
        let ack = Request::<AppOperation<TextApplication>>::decode(
            &client.acknowledgement().unwrap(),
            &binding(),
        )
        .unwrap();
        let Command::Ack {
            original_request_digest,
            result_digest,
        } = ack.command
        else {
            panic!("ack")
        };
        let response = wire::Response::<String, wire::Empty, wire::Empty> {
            schema: wire::RESPONSE_SCHEMA.into(),
            binding: binding(),
            sequence: MAX_ID,
            operation: wire::OperationName::Ack,
            request_digest: ack.request_digest,
            outcome: Outcome::Acknowledged,
            code: wire::Code::ResultReleased,
            body: wire::Body::Acknowledged {
                stamp: wire::AckStamp {
                    sequence: MAX_ID,
                    original_request_digest,
                    result_digest,
                },
            },
            result_digest: String::new(),
        };
        wire::seal_into(response, "result_digest", wire::RESPONSE_SCHEMA, &mut bytes).unwrap();
        client.observe_acknowledgement(&bytes).unwrap();
        assert_eq!(client.next_sequence(), None);
        assert!(client.pending_request().is_none());
        assert!(client
            .begin(AppOperation::<TextApplication>::Prepare(0))
            .is_err());
        // The ordinary initial state still performs a real owner/client exchange.
        let mut owner = Owner::new(
            binding(),
            TextApplication(Rc::new(Cell::new(0))),
            vec!["a".repeat(64)],
        )
        .unwrap();
        let mut client = Client::<TextApplication>::new(binding()).unwrap();
        let request = client
            .begin(AppOperation::<TextApplication>::Prepare(0))
            .unwrap()
            .to_vec();
        client.observe(owner.process(&request).unwrap()).unwrap();
        client
            .observe_acknowledgement(owner.process(&client.acknowledgement().unwrap()).unwrap())
            .unwrap();
        assert_eq!(client.next_sequence(), Some(2));
    }
}
